[Issue 415] Add support for python 3

AndreiH opened the pull request:
This PR is for Issue https://github.com/mozilla/mozdownload/issues/415 to add support for Python 3

AndreiH commented:
@whimboo , can you take a look over this ? 
Thank you !

whimboo commented:
Hi @AndreiH. It's great to see that you pushed a first attempt for the Python3 support! Thanks a lot.

Before I'm going to review this patch we should make sure that at least Travis CI also runs a job for Python3 for both Linux and Mac. This can be done by updating https://github.com/mozilla/mozdownload/blob/master/.travis.yml for the new platforms. 

For appveyor it might also be simple but we would most likely need a matrix in https://github.com/mozilla/mozdownload/blob/master/.appveyor.yml.

whimboo commented:
@AndreiH I have seen that you already modified the travis CI config. Please make sure to keep Python2.7 in the matrix.

AndreiH commented:
Thank you @whimboo , I have updated the matrix as it was before.
It seems that there are some test failures that I need to fix before I can ask for the full review.

whimboo commented:
@AndreiH now you removed Python3 again and only run Python2.7 jobs. What we want is to have both 2.7 and at least 3.6 tests running.

whimboo commented:
Travis CI still doesn't pick up those other jobs. I assume that is because you changed the indentation from 4 to only 2. Make sure to revert it, and it might fix that particular problem.

AndreiH commented:
Hello @whimboo !
I have updated the changes. 
I fixed all tests except one, that I am not really sure why it is failing and I could not find a fix for it.
From debugging and investigating I have reached the conclusion that it's because `treeherder-client` library. Which does not entirely support Python3 yet. BUG -> https://bugzilla.mozilla.org/show_bug.cgi?id=1330474
I see that they have released version 5.0.0, but it still fails.

The single test that is failing is -> https://github.com/AndreiH/mozdownload/blob/master/tests/treeherder/test_api.py#L59

Fails with: 
```_______________________________________________________________________________________ TestAPI.test_query_tinderbox_builds ________________________________________________________________________________________

self = <test_api.TestAPI testMethod=test_query_tinderbox_builds>

    def test_query_tinderbox_builds(self):
        self.httpd.router.register('GET', '/api/*', handle_rest_api)
    
        for platform in PLATFORM_MAP:
            # mac64 is identical to mac
            if platform == 'mac64':
                continue
    
            application = 'firefox' if not platform.startswith('android') else 'mobile'
            th = Treeherder(application, 'mozilla-beta', platform,
                            server_url='http://{}:{}'.format(self.httpd.host, self.httpd.port))
            builds = th.query_builds_by_revision('29258f59e545')
    
>           self.assertEqual(len(builds), 1)
E           AssertionError: 0 != 1

tests/treeherder/test_api.py:72: AssertionError
------------------------------------------------------------------------------------------------ Captured log call -------------------------------------------------------------------------------------------------
client.py                  500 ERROR    HTTPError 500 requesting http://127.0.0.1:8080/api/project/mozilla-beta/jobs/?job_type_name=Build&result_set_id=5319&exclusion_profile=False&build_platform=linux64&option_collection_hash=102210fe594ee9b33d82058545b1ed14f4c8206e: b'{"error": {"code": 500, "message": "Traceback (most recent call last):\\n  File \\"/home/andreihutusoru/mozdownload/.tox/py35/lib/python3.5/site-packages/wptserve/handlers.py\\", line 312, in __call__\\n    rv = self.func(request, response)\\n  File \\"/home/andreihutusoru/mozdownload/.tox/py35/lib/python3.5/site-packages/wptserve/handlers.py\\", line 354, in handle_request\\n    value = enc(rv)\\n  File \\"/usr/lib/python3.5/json/__init__.py\\", line 230, in dumps\\n    return _default_encoder.encode(obj)\\n  File \\"/usr/lib/python3.5/json/encoder.py\\", line 198, in encode\\n    chunks = self.iterencode(o, _one_shot=True)\\n  File \\"/usr/lib/python3.5/json/encoder.py\\", line 256, in iterencode\\n    return _iterencode(o, 0)\\n  File \\"/usr/lib/python3.5/json/encoder.py\\", line 179, in default\\n    raise TypeError(repr(o) + \\" is not JSON serializable\\")\\nTypeError: <filter object at 0x7feedaef2550> is not JSON serializable\\n"}}'
treeherder.py              105 ERROR    Failure occurred when querying Treeherder for builds
Traceback (most recent call last):
  File "/home/andreihutusoru/mozdownload/mozdownload/treeherder.py", line 96, in query_builds_by_revision
    jobs = self.client.get_jobs(self.branch, **kwargs)
  File "/home/andreihutusoru/mozdownload/.tox/py35/lib/python3.5/site-packages/thclient/client.py", line 684, in get_jobs
    return self._get_json_list(self.JOBS_ENDPOINT, project, **params)
  File "/home/andreihutusoru/mozdownload/.tox/py35/lib/python3.5/site-packages/thclient/client.py", line 490, in _get_json_list
    return self._get_json(endpoint, project=project, **params)["results"]
  File "/home/andreihutusoru/mozdownload/.tox/py35/lib/python3.5/site-packages/thclient/client.py", line 497, in _get_json
    resp.raise_for_status()
  File "/home/andreihutusoru/mozdownload/.tox/py35/lib/python3.5/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: http://127.0.0.1:8080/api/project/mozilla-beta/jobs/?job_type_name=Build&result_set_id=5319&exclusion_profile=False&build_platform=linux64&option_collection_hash=102210fe594ee9b33d82058545b1ed14f4c8206e```


I currently xfailed with: `@pytest.mark.skip(reason="Bug 1330474 - Switch Treeherder from Python 2.7 to Python 3.6")`
But I think I need to post another Bug for `treeherder-client`.

Please let me know what you think and if you have any idea how to fix this test.

Thank you !

whimboo commented:
@edmorley mind having a look at the last comment from @AndreiH if the assessment is right? Maybe we need a workaround for now until thclient has been updated.

whimboo commented on .travis.yml:
```diff
@@ -8,17 +8,17 @@ language: python
+          # 10.1 is OS X 10.13.x
           # https://docs.travis-ci.com/user/languages/objective-c/#Supported-OS-X-iOS-SDK-versions
-          osx_image: xcode7.2
+          osx_image: xcode10.1
```
Again. We want both py27 and py36 for Linux and Windows. Means you cannot simply remove the py27 entries.

AndreiH replied:
Ugh, got it !
I will add an new update. Thanks !

AndreiH replied:
But as I see, there is only Linux with Python 2.7 currently.
Do you want me to add also Windows run ?

whimboo replied:
Windows is covered by Appveyor given that Travis CI only supports Linux and MacOS.

whimboo replied:
And the .appveyor config looks all fine and results are all green: https://ci.appveyor.com/project/whimboo/mozdownload/builds/22544085

AndreiH replied:
Got it !

edmorley commented:
> @edmorley mind having a look at the last comment from @AndreiH if the assessment is right? Maybe we need a workaround for now until thclient has been updated.

Hi! 

`treeherder-client` is currently tested under Python 3 in the Treeherder repository and passes our tests (albeit coverage isn't great), as well as passing flake8 and having been checked using pylint's py2 mode.

Looking at the test failure in , I see:

`requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: http://127.0.0.1:8080/api/project/mozilla-beta/jobs/?....`

It looks like the `handle_rest_api` test utility in this repo is generating the HTTP 500, not anything to do with `treeherder-client`? (My hunch would be that the 

Is there a reason that the custom `handle_rest_api()` is being used instead of an existing solutution like [responses](https://github.com/getsentry/responses)?

edmorley commented:
(Comment edited to add the missing parts, whoops)

whimboo commented:
Note that for the pylama job it will be enough to run it with Python3.6

whimboo commented:
> Is there a reason that the custom `handle_rest_api()` is being used instead of an existing solutution like [responses](https://github.com/getsentry/responses)?

No but we haven't had the time yet to work on #400.

edmorley commented:
> No but we haven't had the time yet to work on #400.

Oh I didn't even remember that I had filed that 3 years ago 😆

AndreiH commented:
Thank you for the responses @edmorley  @whimboo !
I have managed to fix the failing test, it was something that I needed to change indeed in `handle_rest_api()` method. It was NOT something to do with `treeherder-client`.
https://github.com/mozilla/mozdownload/pull/536/files#diff-862a7ed7e988689f6ae6678761b9ee14R49
I have never thought of looking into that method. Thanks for the clue.

The PR is now ready for a full review @whimboo !
Travis now has all the running ENVs required, I have changed a bit how we run Travis, as we now have more ENVs.

whimboo commented on .travis/install.sh:
```diff
@@ -0,0 +1,30 @@
+        # install pip on the system python
+        curl -O https://bootstrap.pypa.io/get-pip.py
+        python get-pip.py --user
+        python -m pip install --user virtualenv
```
Why don't you run the `pip` command directly?

AndreiH replied:
I will remove it

whimboo commented on .travis/install.sh:
```diff
@@ -0,0 +1,30 @@
+set -e
+set -x
+
+ci_requirements="pip setuptools tox virtualenv"
```
I don't think that we need pip, and setuptools here, right?

AndreiH replied:
true, let me remove that also.

whimboo commented on .travis/install.sh:
```diff
@@ -0,0 +1,30 @@
+        curl -O https://bootstrap.pypa.io/get-pip.py
+        python get-pip.py --user
+        python -m pip install --user virtualenv
+        python -m virtualenv .venv/
```
I don't see that we need a custom virtualenv. tox will automatically create one for us.

AndreiH replied:
Removed

whimboo commented on .travis/run.sh:
```diff
@@ -0,0 +1,10 @@
+    source .venv/bin/activate
+fi
+
+tox
```
See the other file. Tox creates a virtual env on its own.

AndreiH replied:
I will remove the activation of the venv

whimboo commented on mozdownload/scraper.py:
```diff
@@ -10,22 +10,21 @@
 import redo
 import requests
+from six.moves.urllib.parse import quote
+from six.moves.urllib.parse import urlparse
```
Please combine those lines.

AndreiH replied:
done

whimboo commented on mozdownload/scraper.py:
```diff
@@ -467,13 +468,13 @@ def get_build_info_for_date(self, date, build_index=None):
         parser.entries = parser.filter(self.is_build_dir)
 
-        if has_time:
+        if date.strftime('%H-%M-%S') != '00-00-00':
```
Why that change? Mind explaining it to me?

AndreiH replied:
Sure.
I was going to ask for some help here. I don't know if this is the correct fix.
But multiple `daily_scraper` tests failed with Python 3 with these errors:
https://travis-ci.org/mozilla/mozdownload/jobs/503557491
Something to do with parsing the build time.

whimboo replied:
First we will keep `has_time`, which means that if we need any update in how we check for a valid time, it has to be done where `has_time` gets its value. It would be good if you have a short snippet and run it once with py27 and py36 to see what the difference is.

AndreiH replied:
I updated the `has_time` method

whimboo commented on tests/factory/test_factory_invalid_options.py:
```diff
@@ -2,14 +2,12 @@
 import os
 
 import mozfile
+import mozhttpd_base_test as mhttpd
```
This is a local file so please move it back at the end of the imports.

AndreiH replied:
done

whimboo replied:
Note that those tests have been changed to pytest, and this module is no longer in use. You will have to fix some merge conflicts when rebasing to master. Please do that.

whimboo commented on tests/treeherder/test_api.py:
```diff
@@ -6,23 +6,24 @@
 import pytest
+import six
+from six.moves.urllib.parse import parse_qs
+from six.moves.urllib.parse import urlparse
```
Please combine those two lines.

AndreiH replied:
done

AndreiH commented on .travis.yml:
```diff
@@ -8,32 +8,38 @@ language: python
+          os: osx
+          env:
+            - TOXENV=py36
+            - HOMEBREW_NO_AUTO_UPDATE=1
```
It is strange, but if I do not specify the `HOMEBREW_NO_AUTO_UPDATE=1`, the Travis build with Py3 fails, because of `brew update`.
Here is a job where I removed the `HOMEBREW_NO_AUTO_UPDATE=1` option: https://travis-ci.org/mozilla/mozdownload/jobs/503484085

whimboo replied:
I assume that `brew update` is automatically run? From the log I can see that Python is already installed: `Error: python 3.6.5_1 is already installed`, which comes from `brew install python` in `.travis/install.sh`. Is Python3.6 not installed by default?

AndreiH replied:
I removed the `HOMEBREW_NO_AUTO_UPDATE=1` and the Python install command.
Not it should be more cleaner.

AndreiH commented on .travis.yml:
```diff
@@ -8,32 +8,38 @@ language: python
           env: TOXENV=py27
+        - name: "Run test suite on OS X with Python 3.6"
+          language: generic
+          os: osx
```
Here also, if I add the `osx_image` option, travis fails with the following: https://travis-ci.org/mozilla/mozdownload/builds/503489606?utm_source=github_status&utm_medium=notification

whimboo replied:
Because Python 3.7 is installed and the `python3.6` command cannot be found.

AndreiH replied:
True, I had to specify `osx_image: xcode9.4` for Python 3.6, not `xcode10.1`

whimboo commented on mozdownload/scraper.py:
```diff
@@ -457,7 +457,7 @@ def is_build_dir(self, folder_name):
         """Return the build information for a given date."""
         url = urljoin(self.base_url, self.monthly_build_list_regex)
-        has_time = date and date.time()
+        has_time = date and date.time() and date.strftime('%H-%M-%S') != '00-00-00'
```
Due you know for which value of date this fails? Which test is causing that?

AndreiH replied:
Well, I see in the stacktrace that there are already the `args` and values which the tests fail with: https://travis-ci.org/mozilla/mozdownload/jobs/503557491

whimboo replied:
As best pick one of those failing tests and run it alone by using the -k option like `tox -- -k test_name`. Also specify -s so that you get all the output. I would like to know what `date.time()` differs in between Python2.7 and Python3.

AndreiH replied:
Test `test_build_indices`

Python 3.6: https://gist.github.com/AndreiH/d8cb89dd509a7a4733d58ba706fbdb27
Python 2.7: https://gist.github.com/AndreiH/7b32e6c9a22cfb2212d58ab5d5602769

whimboo replied:
So here is the snippet for your testing: https://gist.github.com/whimboo/63ac15a7029fa2b6be9da33858fb0a91

Note the different behavior between Python2 and Python3. We have to get it fixed, and comparing to `00:00:00` is not a good idea given that at this second we also could have a sub folder.

AndreiH commented on .travis/install.sh:
```diff
@@ -0,0 +1,14 @@
+if [ "$TRAVIS_OS_NAME" == "osx" ]; then
+    curl -O https://bootstrap.pypa.io/get-pip.py
+    python get-pip.py --user
+    pip install --user virtualenv
```
Removed it!

whimboo commented:
@AndreiH will you have a chance to continue on this PR? Note, that there are also conflicts.

AndreiH commented:
> @AndreiH will you have a chance to continue on this PR? Note, that there are also conflicts.

@whimboo Unfortunately I cannot update this on a daily basis, my time is very limited now to work on this.
Hope this is good enough to go ahead with a merge. I know the fix `and date.strftime('%H-%M-%S') != '00-00-00'` is not a good one for now. Maybe next weekend I can take a look. Thanks !

whimboo commented:
We can wait a bit, no problem. There are also remote test failures I would like to see fixed first before merging this PR.

whimboo commented:
> @whimboo Unfortunately I cannot update this on a daily basis, my time is very limited now to work on this.
> Hope this is good enough to go ahead with a merge. I know the fix `and date.strftime('%H-%M-%S') != '00-00-00'` is not a good one for now. Maybe next weekend I can take a look. Thanks !

Yes, let take it for now. It would also fail in Python2 for exactly this time, so that it isn't a regression.

Note that we are going to skip the remote TB tests which cause those failures via #550. Once that landed I think we can also merge this PR. I just want to have a green test suite first.

whimboo commented:
@AndreiH would you mind to rebase your branch against the latest master, and repush? I would like to see how the test failure rate has been improved now with PR #554 landed. Thanks.

whimboo commented:
As just noticed the newly added `packaging` package hadn't been added to the `requirements.txt` file. As such all the release candidate tests were failing. By adding this now both Travis and Appveyor are happy.

whimboo commented:
@AndreiH thanks again for working on this patch! Very much appreciated to see py3 support coming up for the next mozdownload release.
